### PR TITLE
[JENKINS-33507] Add support for the Bitbucket server.

### PR DESCRIPTION
**Pull request by Eernie**
[JENKINS-33507](https://issues.jenkins-ci.org/browse/JENKINS-33507)

I made some changes to the code to enable more support for Bitbucket server.
I also made a plugin for Bitbucket which is currently submitted for approval. When both installed Bitbucket can successfully send events to the jenkins machine. 

See [bitbucket-webhooks-plugin](https://github.com/topicusfinan/bitbucket-webhooks-plugin) repo for more information about the Bitbucket plugin.
[The plugin is also accepted by atlassian.](https://marketplace.atlassian.com/plugins/nl.topicus.bitbucket.bitbucket-webhooks/server/overview)

@reviewbybees

### CODE REVIEW

**Comment by tiny-dancer** [comment]
+1

**Comment by Jurrie** [comment]
+1

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/BitbucketPushEvent.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package com.cloudbees.jenkins.plugins.bitbucket.api;
+
+/**
+ * Represents a push event coming from Bitbucket (webhooks).
+ */
+public interface BitbucketPushEvent
+{
```
Comment (amuniz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/BitbucketPushEvent.java]: Please, follow the current code style in the rest of the code base.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/events/BitbucketCloudPushEvent.java [hunk 2]
```diff
@@ -23,13 +23,14 @@
...
-
+import com.cloudbees.jenkins.plugins.bitbucket.api.BitbucketPushEvent;
 import com.cloudbees.jenkins.plugins.bitbucket.api.BitbucketRepository;
 import com.cloudbees.jenkins.plugins.bitbucket.client.repository.BitbucketCloudRepository;
+import org.codehaus.jackson.annotate.JsonIgnoreProperties;
 
 @JsonIgnoreProperties(ignoreUnknown = true)
-public class BitbucketPushEvent {
+public class BitbucketCloudPushEvent implements BitbucketPushEvent
+{
```
Comment (amuniz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/events/BitbucketCloudPushEvent.java]: Code style

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/hooks/BitbucketSCMSourcePushHookReceiver.java [hunk 3]
```diff
@@ -72,7 +70,18 @@ public HttpResponse doNotify(StaplerRequest req) throws IOException {
             LOGGER.info("Received unknown Bitbucket hook: " + eventKey + ". Skipping.");
             return HttpResponses.error(HttpServletResponse.SC_BAD_REQUEST, "X-Event-Key HTTP header invalid: " + eventKey);
         }
-        type.getProcessor().process(body);
+
+		String bitbucketKey = req.getHeader("X-Bitbucket-Type");
```
Comment (amuniz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/hooks/BitbucketSCMSourcePushHookReceiver.java]: Wrong indentation.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/hooks/BitbucketSCMSourcePushHookReceiver.java [hunk 4]
```diff
@@ -72,7 +70,18 @@ public HttpResponse doNotify(StaplerRequest req) throws IOException {
...
+        BitbucketType instanceType = null;
+        if (bitbucketKey != null) {
+			instanceType = BitbucketType.fromString(bitbucketKey);
```
Comment (amuniz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/hooks/BitbucketSCMSourcePushHookReceiver.java]: Wrong indentation. Are you using spaces or tabs?

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/hooks/BitbucketType.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
+package com.cloudbees.jenkins.plugins.bitbucket.hooks;
+
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+
+public enum BitbucketType
+{
+	CLOUD("cloud"),
+	SERVER("server"),
+	;
```
Comment (amuniz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/hooks/BitbucketType.java]: Wrong code style.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/hooks/BitbucketType.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
+package com.cloudbees.jenkins.plugins.bitbucket.hooks;
+
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+
+public enum BitbucketType
+{
```
Comment (amuniz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/hooks/BitbucketType.java]: Code style.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/hooks/PushHookProcessor.java [hunk 7]
```diff
@@ -23,20 +23,30 @@
...
 
     @Override
-    public void process(String payload) {
+    public void process(String payload, BitbucketType instanceType) {
         if (payload != null) {
             // TODO: generalize this for BB server
-            BitbucketPushEvent push = BitbucketCloudWebhookPayload.pushEventFromPayload(payload);
+            BitbucketPushEvent push;
+            switch (instanceType)
+            {
```
Comment (amuniz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/hooks/PushHookProcessor.java]: Code style.
And why not a plain `if`? It's not going to be more than 2 options.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerWebhookPayload.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import org.codehaus.jackson.map.ObjectMapper;
+
+import java.io.IOException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class BitbucketServerWebhookPayload
+{
```
Comment (amuniz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerWebhookPayload.java]: Code style.

**Comment by amuniz** [comment]
Looks good in general, and I like the idea to use an OSS plugin at Bitbucket side also, thanks for it.

I didn't have time to test it locally, but I want to do it. In the meantime, please fix the code style comments and merge with master (there are conflicts now).

**Comment by Eernie** [comment]
@amuniz Thanks for the review, I've resolved the conflict and fixed the code style.

**Comment by tiny-dancer** [comment]
image
;)

**Comment by nickbroon** [comment]
Is  this now suitable for merging?

**Comment by jreijn** [comment]
I've tested the PR locally in combination with https://marketplace.atlassian.com/plugins/nl.topicus.bitbucket.bitbucket-webhooks/server/overview . Works like a charm.

**Comment by Jurrie** [comment]
@amuniz Can this PR be merged now? It seems your comments are processed.

**Comment by Eernie** [comment]
@amuniz Had any time to test this lately? I think a lot of people are waiting on a new release with this. :)

**Comment by bhat-man** [comment]
Please merge this.  We just purcahses a license of Cloudbees and can't do what we want without this.

**Comment by amuniz** [comment]
@bhat-man the plugin this PR is developed against is [this](https://marketplace.atlassian.com/plugins/nl.topicus.bitbucket.bitbucket-webhooks/server/overview). I think you are using another one.

**Comment by bhat-man** [comment]
I did try that one as well and had similar issues.  I will build a fresh Jenkins server and test again.  Is this the correct endpoint tp use:

/bitbucket-scmsource-hook/notify

**Comment by Mktmac** [comment]
Hi guys,

I was wondering if there's a timetable for the merge of this PR? I saw that everything seems to be working fine.

Thanks,
Dnaiel

**Comment by markottaviani** [comment]
Another Cloudbees customer here, very interested in seeing this get merged if all concerns have been satisfied.  Big thinks to @Eernie for putting this together.

**Comment by aheritier** [comment]
I will try to escalate it internally but I don't expect to have some help this week with Jenkins World. But if some of you are going to JW, don't hesitate to kidnapp @amuniz up until it reviews/merge/release it :-P

**Comment by Eernie** [comment]
@aheritier Thanks for that, I'm hoping the list of PR's can be merged and released quickly. :+1:

**Comment by amuniz** [comment]
@aheritier with friends like you I prefer enemies 😛 
This is going to be merged in a few days (I need to test it against a BB Server, which will take me some time).

BTW we need a way to test this kind of changes ideally without the need of a running BB Server, if someone come up with a solution for that, it would be more than welcome!! (maybe an embedded dumb JSON server returning static files suffice, TBD).

**Comment by aheritier** [comment]
@amuniz 😗 ❤️ I love you too

**Comment by zackpollard** [comment]
Glad to hear this should be merged soon, was trying to use the auto webhook feature earlier today to find out it can only be used with bitbucket cloud! Good work all :)

**Comment by zackpollard** [comment]
@Eernie Is there any reason that this needs to use your own plugin for bitbucket? Could it not use the more widely used (and so more likely to be installed on an enterprise bitbucket) plugin found here https://marketplace.atlassian.com/plugins/com.atlassian.stash.plugin.stash-web-post-receive-hooks-plugin/server/overview

**Comment by Eernie** [comment]
@zackpollard I've tried that first. But after implementing that version I noticed that you're limited to 5 Calls per repo and you're limited to pushes only. The hook wouldn't get triggered when making a pull request. 
And implementing the "auto register" function would be imposible.

**Comment by zackpollard** [comment]
I believe that the pull request thing is actually a bug in the more recent versions of bitbucket. As for auto registering that would be possible as you can do it through the API, I have done this for another automated setup we are currently using.


via Newton Mail [https://cloudmagic.com/k/d/mailapp?ct=pa&cv=8.6.38&pv=7.0&source=email_footer_2] On Thu, Sep 15, 2016 at 9:38 PM, Erwin Oldenkamp < notifications@github.com [notifications@github.com] > wrote:
@zackpollard [https://github.com/zackpollard] I've tried that first. But after implementing that version I noticed that you're limited to 5 Calls per repo and you're limited to pushes only. The hook wouldn't get triggered when making a pull request.
And implementing the "auto register" function would be imposible.

**Comment by Eernie** [comment]
Then still there isn't a way to be as complient as the cloud version is. There you can choose the different events you want to send. The plugin you sent is a close source solution. So creating features for it wouldn't be posible. Therefor implementing the different kind of events wouldn't be possible.
The plugin is not written for this plugin only. We got more plans for it in the future :)

**Comment by zackpollard** [comment]
Yea it would just be nice to have additional options as in large enterprise wide bitbucket instances getting plugins installed can be difficult, so using what is already installed is easier :P


via Newton Mail [https://cloudmagic.com/k/d/mailapp?ct=pa&cv=8.6.38&pv=7.0&source=email_footer_2] On Thu, Sep 15, 2016 at 9:53 PM, Erwin Oldenkamp < notifications@github.com [notifications@github.com] > wrote:
Then still there isn't a way to be as complient as the cloud version is. There you can choose the different events you want to send. The plugin you sent is a close source solution. So creating features for it wouldn't be posible. Therefor implementing the different kind of events wouldn't be possible.
The plugin is not written for this plugin only. We got more plans for it in the future :)

**Comment by amuniz** [comment]
This has been merged into master manually. I fixed some minor glitches (users being uppercase in Bitbucket Server) and other minor changes.

**Pull request merged**